File: training/ruby/spec/de/sots/cellars_and_centaurs/domain/service/creature_service_spec.rb
# frozen_string_literal: true

require 'spec_helper'

require_relative '../../../../../../lib/de/sots/cellars_and_centaurs/domain/service/creature_service'
require_relative '../../../../../../lib/de/sots/cellars_and_centaurs/domain/model/creature'
require_relative '../../../../../../lib/de/sots/cellars_and_centaurs/domain/model/creature_id'
require_relative '../../../../../../lib/de/sots/cellars_and_centaurs/domain/model/speed'

RSpec.describe De::Sots::CellarsAndCentaurs::Domain::Service::CreatureService do
  let(:creatures) { instance_double(De::Sots::CellarsAndCentaurs::Domain::Service::Creatures, save: nil) }
  let(:creature_service) { described_class.new(creatures, Logger.new(nil)) }

  it 'should_save_creature_to_the_stable' do
    # Arrange
    walkingSpeed = De::Sots::CellarsAndCentaurs::Domain::Model::Speed.new(40)
    creature = De::Sots::CellarsAndCentaurs::Domain::Model::Creature.new(
      De::Sots::CellarsAndCentaurs::Domain::Model::CreatureId.new('centaur-1')
    )
    creature.speeds = { walking: walkingSpeed }

    # Act
    creature_service.save(creature)

    # Assert
    expect(creatures).to have_received(:save).with(creature)
  end
end

File: training/vue/src/de/sots/cellarsandcentaurs/domain/service/__tests__/CreatureService.spec.ts
import { describe, expect, it, vi } from 'vitest'
import { CreatureService } from '../CreatureService'
import { Creatures } from '../Creatures'
import { Creature } from '../../model/Creature'
import { CreatureId } from '../../model/CreatureId'
import { Speed } from '../../model/Speed'
import { SpeedType } from '../../model/SpeedType'

describe('CreatureService', () => {
  function should_save_creature_to_the_stable() {
    // Arrange
    const creatures: Creatures = { save: vi.fn(), find: vi.fn() }
    const service = new CreatureService(creatures)
    const walking_speed = new Speed(40)
    const creature = new Creature(new CreatureId('centaur-1'))
    creature.setSpeeds(new Map([[SpeedType.WALKING, walking_speed]]))

    // Act
    service.save(creature)

    // Assert
    expect(creatures.save).toHaveBeenCalledWith(creature)
  }

  it('should_save_creature_to_the_stable', should_save_creature_to_the_stable)
})
